### PR TITLE
core: arm: add support for i.MX6 Dual Lite SabreSD

### PR DESCRIPTION
**Pull request by mbriand**
This patch adds support for the Dual Lite version of the i.MX6 Sabre SD reference board.

In addition to add support for yet another board, it adds support for the Solo/DualLite family of NXP i.MX6 SoCs.

Signed-off-by: Mathieu Briand <mbriand@witekio.com>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`
@MrVan is this OK with you too?

File: core/arch/arm/plat-imx/platform_config.h [hunk 1]
```diff
@@ -128,8 +129,16 @@
 #define GICD_OFFSET			0x1000
 #define GIC_CPU_BASE			(GIC_BASE + GICC_OFFSET)
 #define GIC_DIST_BASE			(GIC_BASE + GICD_OFFSET)
+
+#if defined(PLATFORM_FLAVOR_mx6qsabrelite) || \
+	defined(PLATFORM_FLAVOR_mx6qsabresd)
 #define UART1_BASE			0x02020000
 #define UART2_BASE			0x021E8000
+#else
+#define UART1_BASE			0x02020000
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/platform_config.h]: `UART1_BASE` is the same for the 3 flavors.

File: core/arch/arm/plat-imx/platform_config.h [hunk 2]
```diff
@@ -146,12 +155,22 @@
 #if defined(PLATFORM_FLAVOR_mx6qsabresd)
 #define CONSOLE_UART_BASE		UART1_BASE
 #endif
+#if defined(PLATFORM_FLAVOR_mx6dlsabresd)
+#define CONSOLE_UART_BASE		UART1_BASE
+#define CONSOLE_UART_PA_BASE		UART1_BASE
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/platform_config.h]: `CONSOLE_UART_PA_BASE` is useless.

File: core/arch/arm/plat-imx/platform_config.h [hunk 3]
```diff
@@ -146,12 +155,22 @@
...
+#endif
 #define DRAM0_BASE			0x10000000
 #define DRAM0_SIZE			0x40000000
 
 #define CFG_TEE_RAM_VA_SIZE		(1024 * 1024)
 
+#if defined(PLATFORM_FLAVOR_mx6qsabrelite) || \
+	defined(PLATFORM_FLAVOR_mx6qsabresd)
+#define UART1_BASE			0x02020000
```
Comment (etienne-lms) [inline core/arch/arm/plat-imx/platform_config.h]: `UART1_BASE` is already defined

File: core/arch/arm/plat-imx/platform_config.h (hunk 2 above)
Comment (mbriand) [inline core/arch/arm/plat-imx/platform_config.h]: That's right, sorry, seems to be a merge error during my rebase. I will fix it.

File: core/arch/arm/plat-imx/platform_config.h (hunk 3 above)
Comment (mbriand) [inline core/arch/arm/plat-imx/platform_config.h]: Same as the one above, sorry.

File: core/arch/arm/plat-imx/platform_config.h (hunk 1 above)
Comment (mbriand) [inline core/arch/arm/plat-imx/platform_config.h]: I thought it was better this way, so UART definitions are grouped by SoC family, but I can get UART1_BASE out of the ifdef.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

File: core/arch/arm/plat-imx/platform_config.h (hunk 1 above)
Comment (mbriand) [inline core/arch/arm/plat-imx/platform_config.h]: Actually, I had a look at the datasheet and it appears that:
- UARTs 1, 3 and 5 are at the same addresses on both SoC families.
- UARTs 2 and 4 are only present in the Dual/Quad family.

What do you think about adding all possible UARTs to clarify everything and ease the process to port on any i.MX6 based board ? That would be something like this:

```
#define UART1_BASE          0x02020000                                              
#define UART2_BASE          0x021E8000                                              
#define UART4_BASE          0x021F0000                                              
                                                                                
#if defined(PLATFORM_FLAVOR_mx6qsabrelite) || \                                     
    defined(PLATFORM_FLAVOR_mx6qsabresd)                                            
#define UART3_BASE          0x021EC000                                              
#define UART5_BASE          0x021F4000                                              
#endif 
```

**Comment by MrVan** [comment]
From the long run, we need to make the code easy to maintain and clean. I am moving in a direction such as defined CFG_MX6Q/MX6DL/MX6SX/MX6SL/MX6UL/MX7D for different SoCs. Then for different boards/platforms, we could select CFG_MX[x]. Also introduce imx-regs.h to cover the IPs addresses, but not in platform_config.h. In platform_config.h, use 'include <imx-regs.h>, include <imx6ul.h> and etc' Different SoCs have there own files, such as imx6ul.c/imx7d.c and etc.

Basically I am ok with the patch now. You need to update .travis.yml to avoid build break, after this, you could have `Reviewed-by: Peng Fan <peng.fan@nxp.com>`

**Comment by jforissier** [comment]
> From the long run, we need to make the code easy to maintain and clean. I am moving in a direction such as defined CFG_MX6Q/MX6DL/MX6SX/MX6SL/MX6UL/MX7D for different SoCs. Then for different boards/platforms, we could select CFG_MX[x].

OK.

> Also introduce imx-regs.h to cover the IPs addresses, but not in platform_config.h. In platform_config.h, use 'include <imx-regs.h>, include <imx6ul.h> and etc'

Sounds good!

> Different SoCs have there own files, such as imx6ul.c/imx7d.c and etc.

Yeah, maybe. But please also remember to put code in `drivers/` whenever it makes sense. This will comme handy when we introduce more Device Tree stuff, especially DT-based drivers (which is likely to happen one day or another).

> Basically I am ok with the patch now. You need to update .travis.yml to avoid build break, after this, you could have `Reviewed-by: Peng Fan <peng.fan@nxp.com>`

Thanks @MrVan!

**Comment by mbriand** [comment]
Config added to the  .travis.yml file.

I agree with your comments regarding the complexity of current sources, but I didn't want to add complexity to this patch by doing a code refactor.

I am currently doing some cleaning based on your comments and on some of my previous observations, but I think it should be part of another pull request. Or do you think I should add the patch to this one ?

**Comment by jforissier** [comment]
@mbriand better do the cleanup in a new PR.
Can you please make sure your patch is rebased on top of master, and add the Acked-by/Reviewed-by tags above? Then I can merge it. Thanks!
